src/backend/allocator/format.rs

```diff
@@ -0,0 +1,399 @@
+//! Format info tables for DRM formats.
+//!
+//! This module provides three functions, [`get_opaque`], [`has_alpha`] and [`get_bpp`].
+//!
+//! [`get_opaque`] returns the opaque alternative of a DRM format with an alpha channel.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::get_opaque;
+//! assert_eq!(Some(Fourcc::Xrgb8888), get_opaque(Fourcc::Argb8888));
+//! ```
+//!
+//! [`has_alpha`] returns true if the format has an alpha channel.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::has_alpha;
+//! assert!(has_alpha(Fourcc::Argb8888));
+//! ```
+//!
+//! [`get_bpp`] returns the number of bits per pixel of a format.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::get_bpp;
+//! assert_eq!(get_bpp(Fourcc::Argb8888), Some(32));
+//! ```
+
+/// Macro to generate table lookup functions for formats.
+///
+/// See the module documentation for usage details.
+macro_rules! format_tables {
+    (
+        $($fourcc: ident {
+            $(opaque => $opaque: ident,)?
+            bpp => $bpp: expr,
+        }),*
```

Comment:
Nitpick: let's use `:` to match the struct definition format, and optionally allow a trailing `,` rather than require it. IMO that'd be more practical for future reading and maintenance of this table.

```suggestion
        $($fourcc: ident {
            $(opaque : $opaque: ident,)?
            bpp : $bpp: expr $(,)?
        }),*
```